stan/math/rev/mat/fun/ordered_constrain.hpp

```diff
@@ -0,0 +1,80 @@
+#ifndef STAN_MATH_REV_MAT_FUN_ORDERED_CONSTRAIN_HPP
+#define STAN_MATH_REV_MAT_FUN_ORDERED_CONSTRAIN_HPP
+
+#include <stan/math/prim/arr/err/check_nonzero_size.hpp>
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/rev/mat/fun/adj_jac_apply.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+namespace {
+class ordered_constrain_op {
+  int N_;
+  double* exp_x_;
+
+ public:
+  ordered_constrain_op() : N_(0) {}
+  /**
+   * Return an increasing ordered vector derived from the specified
+   * free vector.  The returned constrained vector will have the
+   * same dimensionality as the specified free vector.
+   *
+   * @param x Free vector of scalars
+   * @return Increasing ordered vector
+   */
+  Eigen::VectorXd operator()(const Eigen::VectorXd& x) {
+    N_ = x.size();
+    exp_x_ = ChainableStack::instance().memalloc_.alloc_array<double>(N_ - 1);
+
+    Eigen::Matrix<double, Eigen::Dynamic, 1> y(N_);
+    if (N_ == 0)
+      return y;
+    y[0] = x[0];
```

Comment:
That's what I meant---I also didn't see a way to share it.  It's annoying when there's just a little bit of injected complexity in an otherwise cut-and-paste algorithm.  It's not worth abstracting out a callback for the guts.